### PR TITLE
Handle pluginPathsService disconnect while loading plugins

### LINKED ISSUE #8219: Reconnect not always triggers activation due to exception
### Bug Description:

On websocket brief disconnections after reconnect there should be activation of all extensions. Without that they stop working.
It seems that not always this happen. Sometime there is this failure:
```
root ERROR Failed to load plugins: Error: Request 'getHostLogPath' failed
    at Proxy.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/bundle.js:134498:33)
    at HostedPluginSupport.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:1012:57)
    at step (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:601:23)
    at Object.next (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:582:53)
    at https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:576:71
    at new Promise (<anonymous>)
    at push.../../packages/plugin-ext/lib/hosted/browser/hosted-plugin.js.__awaiter (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:572:12)
    at HostedPluginSupport.push.../../packages/plugin-ext/lib/hosted/browser/hosted-plugin.js.HostedPluginSupport.startPlugins (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:1002:16)
    at HostedPluginSupport.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:847:51)
    at step (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:601:23)
```
On the backend I see something like this:
```
root ERROR Request getHostLogPath failed with error: Connection got disposed. Error: Connection got disposed.
    at Object.dispose (/theia/node_modules/vscode-jsonrpc/lib/main.js:904:25)
    at connection.onClose (/theia/node_modules/vscode-ws-jsonrpc/lib/socket/connection.js:14:41)
    at CallbackList.invoke (/theia/node_modules/vscode-jsonrpc/lib/events.js:62:39)
    at Emitter.fire (/theia/node_modules/vscode-jsonrpc/lib/events.js:121:36)
    at closeHandler (/theia/node_modules/vscode-jsonrpc/lib/main.js:240:26)
    at CallbackList.invoke (/theia/node_modules/vscode-jsonrpc/lib/events.js:62:39)
    at Emitter.fire (/theia/node_modules/vscode-jsonrpc/lib/events.js:121:36)
    at WebSocketMessageReader.fireClose (/theia/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
    at WebSocketMessageReader.fireClose (/theia/node_modules/vscode-ws-jsonrpc/lib/socket/reader.js:67:19)
    at WebSocketMessageReader.socket.onClose (/theia/node_modules/vscode-ws-jsonrpc/lib/socket/reader.js:24:18)
```

### Steps to Reproduce:

1. Open gitpod with wifi connection
2. Open the theia run there in a separate tab following the notification
3. Install some extensiosn from extension pane
4. Switch between wifi connections
5. Every other time you get the error above in browser console
6. Then no activation is done and extensions do not work anymore

### Additional Information

- Operating System: Gitpod
- Theia Version: master

### PR DESCRIPTION
**Pull request by amiramw**
Fixes #8219 

Sometimes on reconnects start plugins fail with Connection got disposed.

The problematic flow is:
- reconnect happens
- `this.server` is reconnected first and doLoad starts
- `this.pluginPathsService.getHostLogPath()` is called before with old `pluginPathsService` connection
- the `pluginPathsService` on server side computes the response to `getHostLogPath`
- `pluginPathsService` connection got disposed
- the sending of response fail

This can be tested by moving between wifi connections on gitpod.

With this change the error still happens but once this.pluginPathsService connection is restored the loading is triggered again.

Signed-off-by: Amiram Wingarten <amiram.wingarten@sap.com>

#### What it does
In case of exception during loading of plugins it: 
- disposes the half work that was done
- retry once again to load the plugins

#### How to test
- Checkout on gitpod
- Install plugin "Remove empty lines" (but it can be tested with any plugin)
- Switch between wifi connections several times
- After each switch there should be reconnect
- After each switch this should still work: F1 -> Remove empty lines: In document
- After some reconnects there is this error and after that successful reload:
```
root ERROR Failed to load plugins: Error: Request 'getHostLogPath' failed
    at Proxy.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/bundle.js:134498:33)
    at HostedPluginSupport.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:1049:57)
    at step (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:601:23)
    at Object.next (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:582:53)
    at https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:576:71
    at new Promise (<anonymous>)
    at push.../../packages/plugin-ext/lib/hosted/browser/hosted-plugin.js.__awaiter (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:572:12)
    at HostedPluginSupport.push.../../packages/plugin-ext/lib/hosted/browser/hosted-plugin.js.HostedPluginSupport.startPlugins (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:1039:16)
    at HostedPluginSupport.<anonymous> (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:879:51)
    at step (https://3000-af7cdfe6-64f1-4923-bdd8-1fdf8d30e6e7.ws-eu01.gitpod.io/57.bundle.js:601:23)
```

When this error occur before this change it cause all plugins to stop working as they were not reactivated.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/hosted/browser/hosted-plugin.ts [hunk 1]
```diff
@@ -239,50 +239,59 @@ export class HostedPluginSupport {
 
     protected loadQueue: Promise<void> = Promise.resolve(undefined);
     load = debounce(() => this.loadQueue = this.loadQueue.then(async () => {
-        try {
-            await this.progressService.withProgress('', PluginProgressLocation, () => this.doLoad());
-        } catch (e) {
-            console.error('Failed to load plugins:', e);
+        // retry loading if needed due to sporadic "Connection got disposed" error on fast reconnect
```
Comment (akosyakov) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:242]: why do we need it, if there is a reconnection, there will be `onDidOpenConnection` event and `load` is called again from above.

Comment (amiramw) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:242]: All this happens and still the first call to `this.pluginPathsService` which is `getHostLogPath` fail with connection got disposed. The second call which comes 0.001 seconds after succeeds (`getHostStoragePath`).

Comment (amiramw) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:242]: @akosyakov what do you think? do you have any alternative approach? or these "connection got disposed" could happen and then retry is right?

Comment (akosyakov) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:242]: Do you mean that we call different JSON-RPC services and not all can be reconnected already on `onDidOpenConnection` event? I will need to verify it. I thought that it was not the case.

Comment (amiramw) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:242]: From my experience most of the time when switching WiFis this event is triggered but only after few seconds the service is really working. Before that we get connection got disposed.

Does it make sense?

**Comment by amiramw** [comment]
@akosyakov I changed the solution and updated the description to specifically describe the error flow. can you please review again?

**Comment by akosyakov** [comment]
@amiramw Sorry I don't have much time this month. I hope to get more time later this month or beginning next week.

File: packages/plugin-ext/src/hosted/browser/hosted-plugin.ts (hunk 1 above)
Comment (akosyakov) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin.ts]: An idea was that the reconnection should be atomic and it is enough to listen to one json-rpc service to be sure that other are reconnected. Maybe there were some breaking change.

**Comment by amiramw** [comment]
I also can't reproduce it right now, but for sure it reproduced consistently before.

> all channels are disposed and no channel is opened in the same tick.

Is it possible that `this.server` is reconnected on the browser side before the server side realizes that the old socket is closed?

**Comment by amiramw** [comment]
Closing as it does not reproduce anymore.

**Pull request closed without merging**